lib/maths/unittest/CBoostedTreeTest.cc

```diff
@@ -120,6 +119,63 @@ auto predictionStatistics(test::CRandomNumbers& rng,
 }
 }
 
+void CBoostedTreeTest::testPersistRestore() {
+    std::size_t rows{50};
+    std::size_t cols{3};
+    std::size_t capacity{50};
+    test::CRandomNumbers rng;
+
+    std::stringstream persistOnceSStream;
+    std::stringstream persistTwiceSStream;
+
+    TFactoryFunc makeMainMemory{
+        [=] { return core::makeMainStorageDataFrame(cols, capacity).first; }};
+
+    TDoubleVecVec x(cols);
+    for (std::size_t i = 0; i < cols; ++i) {
+        rng.generateUniformSamples(0.0, 10.0, rows, x[i]);
+    }
+    auto frame = makeMainMemory();
+
+    //generate completely random data
+    for (std::size_t i = 0; i < rows; ++i) {
+        frame->writeRow([&](core::CDataFrame::TFloatVecItr column, std::int32_t&) {
+            for (std::size_t j = 0; j < cols; ++j, ++column) {
+                *column = x[j][i];
+            }
+        });
+    }
+    frame->finishWritingRows();
+
+    //persist
+    {
+        auto boostedTree = maths::CBoostedTreeFactory::constructFromParameters(
+                               1, cols - 1, std::make_unique<maths::boosted_tree::CMse>())
+                               .numberFolds(2)
+                               .maximumNumberTrees(2)
+                               .maximumOptimisationRoundsPerHyperparameter(3)
+                               .buildFor(*frame);
+        core::CJsonStatePersistInserter inserter(persistOnceSStream);
+        boostedTree->acceptPersistInserter(inserter);
+        persistOnceSStream.flush();
+    }
+    //restore
+    auto boostedTree =
+        maths::CBoostedTreeFactory::constructFromString(persistOnceSStream, *frame);
+    {
+        core::CJsonStatePersistInserter inserter(persistTwiceSStream);
+        boostedTree->acceptPersistInserter(inserter);
+        persistTwiceSStream.flush();
+    }
+    CPPUNIT_ASSERT_EQUAL(persistOnceSStream.str(), persistTwiceSStream.str());
+    LOG_DEBUG(<< "First string " << persistOnceSStream.str());
+    LOG_DEBUG(<< "Second string " << persistTwiceSStream.str());
+
+    // and even run
+    CPPUNIT_ASSERT_NO_THROW(boostedTree->train());
+    CPPUNIT_ASSERT_NO_THROW(boostedTree->predict());
```

Comment:
A nice test would be that we get the same results after persist and restore vs just running straight through without going through this cycle. I know we're under time pressure to commit this now so maybe we leave this to a follow on change.